### PR TITLE
check POST requests for valid json header

### PR DESCRIPTION
**Pull request by mayya-sharipova**
validate that all POST requests with json body must have also have valid
json header: {"Content-Type": "application/json"}
This ensures a basic protection against CSRF

JIRA: COUCHDB-2775

### CODE REVIEW

File: src/chttpd_db.erl [hunk 1]
```diff
@@ -362,7 +362,6 @@ db_req(#httpd{path_parts=[_,<<"_ensure_full_commit">>]}=Req, _Db) ->
 
 db_req(#httpd{method='POST',path_parts=[_,<<"_bulk_docs">>], user_ctx=Ctx}=Req, Db) ->
     couch_stats:increment_counter([couchdb, httpd, bulk_requests]),
-    couch_httpd:validate_ctype(Req, "application/json"),
```
Comment (kxepal) [inline src/chttpd_db.erl]: Why did you remove the check here?

Comment (mayya-sharipova) [inline src/chttpd_db.erl]: Because after this line follows:  {JsonProps} = chttpd:json_body_obj(Req), and the function json_body_obj already having a check for header.

Comment (kxepal) [inline src/chttpd_db.erl]: mmm...I think it's conceptually wrong: `chttpd:json_body_obj` ensures that request body is a JSON object, but there is more general `json_body` function which need to be insured that request body _is_ a JSON value. 

However, both shouldn't care about mime type - that's not their work and this prevents to reuse these function if request body is a JSON, but mime type is not a `application/json`. For instance, it could be `application/vnd.cloudant+json`, but because of `chttpd:json_body` does additional wrong job, you'll have to create your own function to decode JSON. 

Single responsibility is a good idea to follow.

Comment (mayya-sharipova) [inline src/chttpd_db.erl]: Thanks for the detailed comments, this makes sense. I have updated the commit to reflect them.

**Comment by kxepal** [comment]
LGFM. Did you check other apps which provides chttpd endpoints for POST content-type validation?

**Comment by mayya-sharipova** [comment]
Thanks for the comments. Not sure what LGFM means. For the other apps: I plan to validate headers for POST requests also in couch, couch_mrview and mango.

**Comment by kxepal** [comment]
@mayya-sharipova Looks Good For Me (:
@rnewson +1?

**Comment by rnewson** [comment]
I'd rather this was chttpd:validate_ctype instead, but it looks good otherwise. Let's add validate_ctype to chttpd and have that function, and only that one, call down to couch_httpd.

**Comment by rnewson** [comment]
+1
